### PR TITLE
Generalize POI to request arbitrary tags

### PR DESCRIPTION
**Pull request by kylebarron**
Similar to the closed PR #240. 

Functions are modified to accept a `tags` argument, which is a dictionary of different key-value pairs. Similar to #240, values of the `dict` can be `True`, `str`, or list of `str`. I.e.
```
tags = {
    'amenity':True,
    'landuse':['retail','commercial'],
    'highway':'bus_stop'
}
```
I didn't try to implement `key=False`, to remove any items of `key` from the search.

In contrast to #240, this leaves the `amenity` parameter in the POI functions, so it doesn't break backwards compatibility. If both `amenities` and `tags['amenity']` are supplied, then they are combined before the Overpass request.

```py
tags = {'tourism': 'motel'}
west = -118.455619
south = 37.330447
east = -118.380432
north = 37.390072
test = create_poi_gdf(north=north, south=south, west=west, east=east, tags=tags)
print(test[['osmid', 'geometry', 'name', 'tourism']].head(5))
```
```
                 osmid                     geometry                     name  \
1925945217  1925945217  POINT (-118.39579 37.36566)          Townhouse Motel   
1925945246  1925945246  POINT (-118.39458 37.36557)       Econo Lodge Bishop   
1925945270  1925945270  POINT (-118.39414 37.36592)        Travelodge Bishop   
1925945363  1925945363  POINT (-118.39573 37.35655)           Super 8 Bishop   
2934354040  2934354040  POINT (-118.39274 37.36029)  Americas Best Value Inn   

           tourism  
1925945217   motel  
1925945246   motel  
1925945270   motel  
1925945363   motel  
2934354040   motel  
```

### CODE REVIEW

**Comment by kylebarron** [comment]
The CI fail doesn't seem related to my changes

**Comment by kylebarron** [comment]
I think this CI fail is the same as reported in #347

**Comment by gboeing** [comment]
@kylebarron CI should be fixed in #350

**Comment by kylebarron** [comment]
Great. I can add some tests to fix the coverage fail, but first do you have comments on the changes themselves? For this and #343? I don't want to write the tests before needing to change parts of the PR

File: osmnx/pois.py [hunk 1]
```diff
@@ -76,32 +157,51 @@ def osm_poi_download(polygon=None, amenities=None, north=None, south=None, east=
...
         west, south, east, north = polygon.bounds
 
-        # Parse the Overpass QL query
-        query = parse_poi_query(amenities=amenities, west=west, south=south, east=east, north=north)
-
-    elif not (north is None or south is None or east is None or west is None):
-        # TODO: Add functionality for subdividing search area geometry based on max_query_area_size
-        # Parse Polygon from bbox
-        #polygon = bbox_to_poly(north=north, south=south, east=east, west=west)
+    elif not all(isinstance(x, float) for x in [north, south, east, west]):
```
Comment (gboeing) [inline osmnx/pois.py]: Will this fail if the passed in bounding box contains integer coordinates? They should be valid too.

File: osmnx/pois.py [hunk 2]
```diff
@@ -37,36 +38,116 @@ def parse_poi_query(north, south, east, west, amenities=None, timeout=180, maxsi
...
-        # Parse amenties
-        query_str = query_template.format(north=north, south=south, east=east, west=west,
-                                          timeout=timeout, maxsize=maxsize)
-
-    return query_str
-
-
-def osm_poi_download(polygon=None, amenities=None, north=None, south=None, east=None, west=None,
-                     timeout=180, max_query_area_size=50*1000*50*1000):
+    # Convert input into dict of tags
```
Comment (gboeing) [inline osmnx/pois.py]: Lines 68-98 could use some more comments. First at the top it'd be useful to document more clearly what this chunk of code does and why. Second, throughout this chunk of code it'd be useful to comment more clearly what the line's logic is. It's a little opaque right now.

File: osmnx/pois.py [hunk 3]
```diff
@@ -37,36 +38,116 @@ def parse_poi_query(north, south, east, west, amenities=None, timeout=180, maxsi
...
+            # The `(._; >;);` joins the new result set into the previous result
+            # set: https://wiki.openstreetmap.org/wiki/Overpass_API/Overpass_QL#Item
+            tag_str += '];(._;>;);'
+
+            for t in ['node', 'way', 'relation']:
+                query.append('({t}{tag_str});'.format(t=t, tag_str=tag_str))
+
+    query.append(');')
+    query.append('out;')
+    return ''.join(query)
```
Comment (gboeing) [inline osmnx/pois.py]: One thing that may be useful: could you provide an example returned value as a commented demonstration? What does the resulting string look like at the end?

**Comment by gboeing** [comment]
The code overall looks good. I left a few comments to make it a bit more intuitive to review. I'll check the results of the tests/coverage when you're ready for me to.

File: osmnx/pois.py (hunk 1 above)
Comment (kylebarron) [reply osmnx/pois.py]: Yes, good catch.

File: osmnx/pois.py (hunk 2 above)
Comment (kylebarron) [reply osmnx/pois.py]: I think I added a lot of comments, and simplified the code a bit as well, at the risk of being too verbose.

**Comment by kylebarron** [comment]
It might also be helpful to be clearer about which function a user is actually intended to use. I didn't actually realize until right now that there are `pois_from_{point,address,bbox,polygon,place}` functions in `pois.py`, and those are presumably the user-oriented functions. When you read through the [`osmnx.pois` docs](https://osmnx.readthedocs.io/en/stable/osmnx.html#module-osmnx.pois), you see all the inner functions and it's hard to notice that the `pois_from_*` functions exist.

**Comment by kylebarron** [comment]
Given this test https://github.com/gboeing/osmnx/blob/c16f10b26e2066fb474da4dbec51599dcf35660b/tests/test_osmnx.py#L360
it looks like historically if no parameters were provided, it was akin to the current `tags={'amenity':True}`. Right now I require either `amenity` or `tags` to be not `None`, but I could switch to the former if you wanted.

Also just a comment, I'm used to running a formatter like [yapf](https://github.com/google/yapf), [black](https://black.readthedocs.io/en/stable/), and/or [isort](https://github.com/timothycrosley/isort) through my code to keep all the code style consistent, especially when working with other people. There are some portions of the code with really long lines and generally inconsistent style that could be helped by using one of these tools.

**Comment by gboeing** [comment]
Circling back around on outstanding PRs now that the semester is over and I can merge them... @kylebarron are you interested in getting this wrapped up and merged? Looks like it's nearly done.

> Right now I require either amenity or tags to be not None, but I could switch to the former if you wanted.

I think that change would be for the best.

> I'd support prefixing each of the non pois_from_* functions with _, which would remove them from the docs pages, though that would break backwards compatibility.

This needs to be done throughout the package, and will be a bigger refactor. Let's leave it alone for this specific feature for now.

**Comment by kylebarron** [comment]
I think it was very close to finished, but I haven't looked at it in 5 months, and I don't use `osmnx` anymore.

**Comment by gboeing** [comment]
@kylebarron ok, no prob.

**Comment by gboeing** [comment]
superseded by #449

**Comment by AnoopDusadh0406** [comment]
@gboeing sir I want to find out list of hospital and  hostel( name & address) for a given city . I tried pois function and geometries_from_place and use tag =  {'Hospital': True}  but did not find any solution . can you help me for finding solution of these type queries. Thanks advanced.

**Pull request closed without merging**